### PR TITLE
demo: leaff demo

### PR DESCRIPTION
**Pull request by alexjbest**
Using #9382 as a somewhat random example

---

Demo for #8479.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by adri326** [comment]
Oh, that's quite neat!

**Comment by eric-wieser** [comment]
I think it would be helpful to group by file there, rather than repeating the `@@ Mathlib.LinearAlgebra.Finsupp @@` heading three times.

**Comment by alexjbest** [comment]
> I think it would be helpful to group by file there, rather than repeating the `@@ Mathlib.LinearAlgebra.Finsupp @@` heading three times.

Yeah there are two competing heuristics here, group by file or put "more important changes first". I opted for the latter as it seemed more helpful to me for long diffs, but if there was consensus for the other one I don't mind changing. Some PRs generate a lot of proof changes though which can be spammy otherwise

**Comment by eric-wieser** [comment]
The one in particular that confused me was the `attribute instance added to LinearMap.CompatibleSMul.finsupp_dom` not being listed right next to `added LinearMap.CompatibleSMul.finsupp_dom`, so maybe my request was an x/y problem.

**Comment by alexjbest** [comment]
So currently there is a hand written list of priorities which controls order diffs are shown in https://github.com/alexjbest/leaff/blob/9dea5f69fb27cad88cfeef57b6af8489f8b40176/Leaff/Diff.lean#L189C22-L190C1
Things are ordered by priority then grouped by module. At the moment I made all priorities distinct, but I can definitely see the argument for changing that so that things like adds and removals and attributes still come first (grouped by module), and more boring stuff later, but with less new module headers the whole time than there are right now

**Comment by eric-wieser** [comment]
I think I might be asking for something more basic like "don't list additions of attributes if the declaration itself is new"

File: .github/workflows/bors.yml [hunk 1]
```diff
@@ -318,3 +318,88 @@ jobs:
...
+          cd leaff/
+          lake build
+
+      - name: run leaff
+        run: |
+          cd leaff/
+          echo "https://github.com/alexjbest/leaff reports" > leaff.diff
+          echo "\`\`\`diff" >> leaff.diff
+          ./runleaff.sh Mathlib ../old/ ../new/ >> leaff.diff
+          echo "\`\`\`" >> leaff.diff
```
Comment (eric-wieser) [inline .github/workflows/bors.yml:396]: A `<details>`/`<summary>` tag along the line of

<details>

<summary>37 changes</summary>

```diff
+This
```
</details>

might be nice for making this use less vertical space

Comment (alexjbest) [reply .github/workflows/bors.yml:396]: Yeah I'd like to implement this, but maybe as 6 additions 7 deletions, 8 attribute additions all as separate sections? What do you think?

**Comment by alexjbest** [comment]
> I think I might be asking for something more basic like "don't list additions of attributes if the declaration itself is new"

This is another design decision I was indeed unsure of, I already implemented the reverse direction of not listing attribute removals when the decl is deleted. But I'm not yet sure what is most useful. For example if your proof breaks at a simp call, it's very helpful for the diff tool to tell you clearly which simp lemmas were added or removed without you having to think further

**Comment by joneugster** [comment]
@alexjbest am I right that some form of this feature meanwhile has been implemented in CI and therefore this PR can be closed?

Please reopen it if I missunderstood! (for example as draft or "WIP" if not ready for review)

**Comment by alexjbest** [comment]
Well the pr was a demo of something that has not been merged as far as I know. Closing this makes sense #8479 was the actual pr

**Pull request closed without merging**